pkg/agent/openflow/network_policy.go

```diff
@@ -184,100 +212,191 @@ type conjMatchFlowContext struct {
-		if err != nil {
-			return err
-		}
+		return ctx.createOrUpdateConjunctiveMatchFlow(actions)
 	}
-	delete(ctx.actions, conjID)
-	return nil
 }
 
-// addAction adds the specified conjunction into conjunctiveMatchFlow's actions, and then updates the conjunctive
-// match flow on the switch. It also installs default drop flow if dropTable is not nil, and the dropFlow is not
-// installed before.
-func (ctx *conjMatchFlowContext) addAction(action *conjunctiveAction) error {
-	// Check if the conjunction exists in conjMatchFlowContext actions or not. If yes, return nil directly.
-	// Otherwise, add the new action, and update the Openflow entry.
+// addAction adds the specified policyRuleConjunction into conjunctiveMatchFlow's actions, and then returns the flowChange.
+func (ctx *conjMatchFlowContext) addAction(action *conjunctiveAction) *flowChange {
+	// Check if the conjunction exists in conjMatchFlowContext actions or not. If yes, return nil immediately.
 	_, found := ctx.actions[action.conjID]
 	if found {
 		return nil
 	}
 
-	// Install or update Openflow entry for the new conjunctiveAction.
-	actions := make([]*conjunctiveAction, 0, len(ctx.actions)+1)
+	// Append current conjunctive action to the existing actions, and then calculate the conjunctive match flow changes.
+	actions := []*conjunctiveAction{action}
 	for _, act := range ctx.actions {
 		actions = append(actions, act)
 	}
-	actions = append(actions, action)
-	err := ctx.installOrUpdateFlow(actions)
-	if err != nil {
-		return err
-	}
-	ctx.actions[action.conjID] = action
-
-	return nil
+	return ctx.createOrUpdateConjunctiveMatchFlow(actions)
 }
 
-func (ctx *conjMatchFlowContext) addDenyAllRule(ruleID uint32) error {
+func (ctx *conjMatchFlowContext) addDenyAllRule(ruleID uint32) {
 	if ctx.denyAllRules == nil {
 		ctx.denyAllRules = make(map[uint32]bool)
 	}
 	ctx.denyAllRules[ruleID] = true
-	return nil
+}
+
+func (ctx *conjMatchFlowContext) delDenyAllRule(ruleID uint32) {
+	// Delete the DENY-ALL rule if it is in context denyAllRules.
+	_, found := ctx.denyAllRules[ruleID]
+	if found {
+		delete(ctx.denyAllRules, ruleID)
+	}
+}
+
+// conjMatchFlowContextChange describes the changes of a conjMatchFlowContext. It is generated when a policyRuleConjunction
+// is added, deleted, or the addresses in an existing policyRuleConjunction are changed. The changes are calculated firstly,
+// and then applied on the OVS bridge using a single Bundle, and then used to update the local cache at last. The local cahce
+// is updated only if conjMatchFlowContextChange is applied on the OVS bridge successfully.
+type conjMatchFlowContextChange struct {
+	// context is the changed conjMatchFlowContext, which needs to be updated after the OpenFlow entries are applied to
+	// the OVS bridge. context is not nil.
+	context *conjMatchFlowContext
+	// ctxChangeType is the changed type of the conjMatchFlowContext. Its value includes "insertion", "modify" and "deletion". Add the
```

Comment:
s/Its value includes/The possible values are

Reply:
Done.